lib/view_component/base.rb

```diff
@@ -161,16 +167,47 @@ def with_variant(variant)
       self
     end
 
+    def with_content(value = nil, &block)
+      if value && block
+        raise ArgumentError.new("Content provided in two ways, using both an argument and a block. Use one or the other.")
+      elsif value.nil? && block.nil?
+        raise ArgumentError.new("No content provided. Provide as an argument or a block.")
+      elsif block
+        @content_set_by_with_content = block
+      else
+        @content_set_by_with_content = -> (_component) { value }
+      end
+
+      self
+    end
+
+    def content
+      return @content if defined?(@content)
+      return unless defined?(@content_set_by_with_content)
+
+      @content = render_or_return(@content_set_by_with_content.call(self))
+    end
+
     private
 
+    # Renders the given object and returns the result, if object can be rendered.
+    # Otherwise, returns object.
+    def render_or_return(object)
+      if object.respond_to?(:render_in)
```

Comment:
I'm not 100% sure, we may be able to avoid this check?

Each component implements `render` that *should* have the correct data available by the time the `with_content` block is eval'd. I wonder if the following would work and also solve the same use-cases?

```ruby
MyComponent.new.with_content do |c|
  c.render MyButtonComponent.new(text: "Submit")
end
```

Reply:
Yes, you're right: the block form allows you to call render yourself and was my initial approach.

After some use, I found myself trying to use components as content, just like strings. When the framework knows how to handle it, similar to how `ActionView` knows to escape a string or not, it is possible to build a view of components concisely.

I'm on mobile and can't provide a code example but would be happy to dive into this further tomorrow. M